### PR TITLE
feat(maven-wrapper): Add manager to update maven wrapper version

### LINKED ISSUE #3566: support maven-wrapper like gradle-wrapper
**What would you like Renovate to be able to do?**
A lot of new projects which are using maven relay on the maven-wrapper (https://github.com/takari/maven-wrapper) so maven behaves in the same way like gradle. you don't need to install a maven locally, it's just calling `./mvnw` and it will download the defined version in `maven-wrapper.properties`. So it would be really nice if there is also support for updating maven-wrapper. Maven wrapper consists of 4 files, 2 of them are stored in `.mvn/wrapper/` directory the other two are the executables in the project root directory

- ./mvn/wrapper/maven-wrapper.jar - downloading logic
- ./mvn/wrapper/maven-wrapper.properties - contains the url and version of maven which should be used for the project
- ./mvnw - linux executable
- ./mvnw.cmd - windows executable 

**Describe the solution you'd like**
solution should consist of two parts:
- detection if new maven is available and update the version for maven just in the properties file
- detection if new wrapper is available und update possibly all 4 files (exectables, properties file and download logic)

**Describe alternatives you've considered**
first thing can be achieved by manually changing the maven version number in maven-wrapper.properties
second thing can be also achieved by manually executing `mvn -N io.takari:maven:0.7.6:wrapper -Dmaven=3.3.3` where the version of maven is pinned to the currently used on and a new version of the wrapper is specified

**Additional context**
maybe the logic from the gradle-wrapper check can be reused and extended to cover both situations for maven (gradle uses only one version for the wrapper and the distribution)

### PR DESCRIPTION
**Pull request by benkelaar**
Merge request for Issue #3566 , we still need to add some unit tests to get 100% coverage and look at whether the documentation needs to be updated, but we would like some first feedback on whether we've taken the right approach.

## Changes
Add custom manager to do maven wrapper version upgrades of both maven version and maven wrapper version.

## Context
- Closes #3566

## Documentation (please check one with an [x])
- [ ] I have updated the documentation, or (WIP)
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [X] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/maven-wrapper/readme.md
Comment (rarkins) [inline lib/modules/manager/maven-wrapper/readme.md:1]:
```suggestion
Configuration for Maven Wrapper updates.
Changes here affect how Renovate updates the version of Maven in the wrapper, not how it uses the wrapper.
```

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/modules/manager/maven-wrapper/readme.md
Comment (benkelaar) [reply lib/modules/manager/maven-wrapper/readme.md]: Thanks, good catch!

File: lib/modules/manager/maven-wrapper/__fixtures__/only-maven/maven-wrapper.properties
Comment (viceice) [inline lib/modules/manager/maven-wrapper/__fixtures__/only-maven/maven-wrapper.properties:1]: please inline those small fixtures

File: lib/modules/manager/maven-wrapper/types.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/types.ts:1]:
```suggestion
export interface Version {
```

prefer interfaces

File: lib/modules/manager/maven-wrapper/index.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/index.ts:8]:
```suggestion
  fileMatch: ['(^|\\/).mvn/wrapper/maven-wrapper.properties$'],
```

escape for safty

File: lib/modules/manager/maven-wrapper/extract.spec.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/extract.spec.ts:2]:
```suggestion
import { extractPackageFile } from '.';
```

better code coverage

File: lib/modules/manager/maven-wrapper/extract.spec.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/extract.spec.ts:1]: https://github.com/renovatebot/renovate/blob/main/docs/development/best-practices.md#fixtures

File: lib/modules/manager/maven-wrapper/artifacts.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/artifacts.ts:44]:
```suggestion
```

Not used

File: lib/modules/manager/maven-wrapper/artifacts.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/artifacts.ts:135]: i think we need semver here

File: lib/modules/manager/maven-wrapper/artifacts.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/artifacts.ts:31]:
```suggestion
```

not required here

File: lib/modules/manager/maven-wrapper/artifacts.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/artifacts.ts:38]: there is no java 5, we only have 8+, so we can't use older maven than 3.1

**Review by viceice — CHANGES_REQUESTED** [review]
needs tests

File: lib/modules/manager/maven-wrapper/extract.spec.ts
Comment (trystian1) [reply lib/modules/manager/maven-wrapper/extract.spec.ts:1]: thanks will implement

File: lib/modules/manager/maven-wrapper/__fixtures__/only-maven/maven-wrapper.properties
Comment (trystian1) [reply lib/modules/manager/maven-wrapper/__fixtures__/only-maven/maven-wrapper.properties:1]: will do

Comment (viceice) [reply lib/modules/manager/maven-wrapper/__fixtures__/only-maven/maven-wrapper.properties:1]: please reduce noise, checkout https://github.com/renovatebot/renovate/blob/main/.github/contributing.md#resolve-review-comments-instead-of-commenting

File: docs/development/best-practices.md
Comment (viceice) [inline docs/development/best-practices.md:225]: please make this in a separate PR

File: lib/modules/manager/maven-wrapper/extract.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/extract.ts:46]:
```suggestion
        depName: 'maven',
        packageName:  'org.apache.maven:apache-maven',
```

File: lib/modules/manager/maven-wrapper/index.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/index.ts:9]:
```suggestion
  versioning,
```
wrong property name

File: lib/modules/manager/maven-wrapper/index.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/index.ts:2]:
```suggestion
import { id as versioning } from '../../versioning/maven';
```

File: lib/modules/manager/maven-wrapper/extract.ts
Comment (trystian1) [reply lib/modules/manager/maven-wrapper/extract.ts:60]: Maven and maven-wrapper can have different versions right? But maybe I understand it wrong.

```
distributionUrl=https://artifactory.tools.bol.com/artifactory/maven-bol/org/apache/maven/apache-maven/3.8.4/apache-maven-3.8.4-bin.zip
wrapperUrl=https://artifactory.tools.bol.com/artifactory/maven-bol/org/apache/maven/wrapper/maven-wrapper/3.1.0/maven-wrapper-3.1.0.jar
```

Comment (viceice) [reply lib/modules/manager/maven-wrapper/extract.ts:60]: ok, thanks for confiming

File: lib/modules/manager/maven-wrapper/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/artifacts.spec.ts:7]:
```suggestion
import { updateArtifacts } from '.';
```

File: lib/modules/manager/maven-wrapper/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/artifacts.spec.ts:1]: mock execution like in gradle wrapper

https://github.com/renovatebot/renovate/blob/37e5d01e0427b2cc72b36be4b921d66a3b31c0da/lib/modules/manager/gradle-wrapper/artifacts.spec.ts#L4

File: lib/modules/manager/maven-wrapper/__fixtures__/wrapper-and-maven/maven-wrapper.properties
Comment (viceice) [inline lib/modules/manager/maven-wrapper/__fixtures__/wrapper-and-maven/maven-wrapper.properties:1]: inline this small fixture

File: lib/modules/manager/maven-wrapper/extract.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/extract.ts:24]: no `any`, create an interface.

File: lib/modules/manager/maven-wrapper/extract.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/extract.ts:36]: use a `for...of` loop and fallback to return `null`

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/manager/maven-wrapper/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/maven-wrapper/readme.md:2]: From reading the docs, it's not clear why our users should care about any of this? It feels we're missing an explanation of the _why_ behind this maven-wrapper thing. 😉

I was expecting something like this:

> By default, Renovate updates Maven in the wrapper like [explain update method].
> This means that [insert problem with current method].
>
> To get around these problem(s), we have a Maven Wrapper manager, which does:
> 
> 1. text
> 2. text
> 3. text

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
The docs should explain why we're even creating a Maven Wrapper manager in the first place. 😉

**Comment by lfvjimisola** [comment]
Any chance of this PR being accepted? Would be very useful for us.

File: lib/modules/manager/maven-wrapper/readme.md
Comment (rarkins) [reply lib/modules/manager/maven-wrapper/readme.md:2]: It's modelled on https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/gradle-wrapper/readme.md

**Comment by rarkins** [comment]
There have been some changes requested during the last review by @viceice. If these have been addressed, please request a re-review by him.

**Comment by trystian1** [comment]
> There have been some changes requested during the last review by @viceice. If these have been addressed, please request a re-review by him.

I think I fixed the review comments, but I cannot ask for a re-review or resolve them

File: lib/modules/manager/maven-wrapper/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/artifacts.spec.ts:171]: please also validate that no command is executed via empty exec snapshots

File: lib/modules/manager/maven-wrapper/artifacts.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/artifacts.ts:137]: wrong, checkout Gradle wrapper manager

https://github.com/renovatebot/renovate/blob/4c9efa4e284807c55143a7c91c5b4f685a4ff178/lib/modules/manager/gradle-wrapper/artifacts.ts#L150-L160

File: lib/modules/manager/maven-wrapper/artifacts.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/artifacts.ts:153]: shouldn't we return an artifact error, so the user is aware of a failure?

**Review by viceice — CHANGES_REQUESTED** [review]
needs docker and install mode tests. check Gradle wrapper manager

File: lib/modules/manager/maven-wrapper/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/artifacts.spec.ts:55]:
```suggestion
    mockedFunction(getPkgReleases).mockResolvedValueOnce({
```

File: lib/modules/manager/maven-wrapper/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/artifacts.spec.ts:47]: please add a comment for which file is read here.

why is it needed at all?

File: lib/modules/manager/maven-wrapper/extract.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/extract.ts:67]:
```suggestion
  return deps.length ? { deps } : null;
```
return null if no deps are extracted

File: lib/modules/manager/maven-wrapper/artifacts.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/artifacts.ts:187]: add a check for os, should not done on windows.

see latest Gradle wrapper changes

File: lib/modules/manager/maven-wrapper/artifacts.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/artifacts.ts:189]: log a warning like the gradle wrapper manager

File: lib/modules/manager/maven-wrapper/artifacts.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/artifacts.ts:157]: it should use relative path.

you need to set appropriate cwdFile to exec command, so renovate will automatically switch to right folder.

File: lib/modules/manager/maven-wrapper/artifacts.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/artifacts.ts:100]: remove null

File: lib/modules/manager/maven-wrapper/artifacts.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/artifacts.ts:102]: filter null values before adding to result

File: lib/modules/manager/maven-wrapper/artifacts.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/artifacts.ts:132]: add cwdFile so we are in right folder for execute

File: lib/modules/manager/maven-wrapper/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/artifacts.spec.ts:234]: use mockExecAll and pass the error instead

**Review by rarkins — COMMENTED** [review]
Any public demo repos available? I'd like to see it demonstrated both in the root of a repo as well as subfolder, assuming that's supported.

What would this require preinstalled to work inside a VM? e.g. the JRE only, or JRE + Maven? I'm guessing `mvnw` will itself pull/install Maven as needed, but it will rely on Java being preinstalled?

**Comment by viceice** [comment]
> Any public demo repos available? I'd like to see it demonstrated both in the root of a repo as well as subfolder, assuming that's supported.
> 
> What would this require preinstalled to work inside a VM? e.g. the JRE only, or JRE + Maven? I'm guessing `mvnw` will itself pull/install Maven as needed, but it will rely on Java being preinstalled?

Yes, only java needs to be installed, like gradle wrapper.

File: lib/modules/manager/maven-wrapper/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/artifacts.spec.ts:174]: please format like this:
https://github.com/renovatebot/renovate/blob/07460542850cc79e934a8d4606371c88776e4c0c/lib/modules/manager/gradle-wrapper/artifacts.spec.ts#L201-L221

please also add `binarySource: 'install'` test

File: lib/modules/manager/maven-wrapper/artifacts.ts
Comment (viceice) [reply lib/modules/manager/maven-wrapper/artifacts.ts:103]: please filter inside loop

File: lib/modules/manager/maven-wrapper/artifacts.ts
Comment (viceice) [reply lib/modules/manager/maven-wrapper/artifacts.ts:101]:
```suggestion
  const updatedResults: UpdateArtifactsResult[] = [];
```

File: lib/modules/manager/maven-wrapper/artifacts.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/artifacts.ts:121]: are you sure maven 3.0.0 already supports java 17? It was released at `2010-10-08` so i'm pretty sure it's not. Please validate with real repo tests

File: lib/modules/manager/maven-wrapper/artifacts.ts
Comment (viceice) [reply lib/modules/manager/maven-wrapper/artifacts.ts:133]: nope, still missing. see
https://github.com/renovatebot/renovate/blob/07460542850cc79e934a8d4606371c88776e4c0c/lib/modules/manager/gradle-wrapper/artifacts.ts#L149

**Comment by benkelaar** [comment]
@rarkins Good point, we tested it on our company's dedicated gitlab instance, will see about setting up a test repo on github.

File: lib/modules/manager/maven-wrapper/artifacts.ts
Comment (benkelaar) [reply lib/modules/manager/maven-wrapper/artifacts.ts:121]: The support should mostly goes the other way right?  Java should be backwards compatible enough to run maven 3.0.0, but will test it in practice.  Good feedback, thanks.

**Comment by rarkins** [comment]
Hi @benkelaar @trystian1 can you use the "re-request reviews" feature in the GitHub web UI to alert reviewers when their comments have been addressed and you want them to take another look? FYI we often _won't_ re-review just because new commits have been made, if there are unresolved conversations and/or no request for a new review

**Comment by rarkins** [comment]
If the previous PR comments have been addressed, please use the "request reviews" option in Github so we know:

image

**Comment by trystian1** [comment]
@benkelaar and me created this demo repo, which demonstrates updating maven wrapper and maven.
https://github.com/benkelaar/demo-repository-renovate-3566/pulls

This was run locally with `yarn start /path-to-repo` 

Any other scenario's you want us to map out?

**Comment by rarkins** [comment]
The example repo looks good to me

**Review by rarkins — DISMISSED** [review]

File: lib/modules/manager/maven-wrapper/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/artifacts.spec.ts:58]: store result and expect `toBeEmtpyArray()`

File: lib/modules/manager/maven-wrapper/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/artifacts.spec.ts:121]: same here

File: lib/modules/manager/maven-wrapper/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/artifacts.spec.ts:163]: plesse format like other managers like here:
https://github.com/renovatebot/renovate/blob/a3587fdfcf0d54623de14fe07fc3505463a020a7/lib/modules/manager/gomod/artifacts.spec.ts#L296-L315

File: lib/modules/manager/maven-wrapper/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/artifacts.spec.ts:27]: please set like here:
https://github.com/renovatebot/renovate/blob/a3587fdfcf0d54623de14fe07fc3505463a020a7/lib/modules/manager/gomod/artifacts.spec.ts#L47-L52

so we're more close to real run paths

File: lib/modules/manager/maven-wrapper/artifacts.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/artifacts.ts:19]: convert to interface

File: lib/modules/manager/maven-wrapper/artifacts.ts
Comment (viceice) [reply lib/modules/manager/maven-wrapper/artifacts.ts:126]: did you validated this?

**Comment by candrews** [comment]
I'm still really hoping to see this merged :crossed_fingers: 

Thank you all for this great work and your continued efforts!

**Comment by viceice** [comment]
some small issues need to be fixed before merge

File: lib/modules/manager/maven-wrapper/artifacts.ts
Comment (benkelaar) [reply lib/modules/manager/maven-wrapper/artifacts.ts:126]: Yes, we tested it both locally and in repositories in github & gitlab.

There is no problem with java backwards compatibility.

However, there are some newer versions of the wrapper that do not work with older versions of maven, but we don't think that's a problem since you can always manually upgrade both to fix this. Is that OK?

File: lib/modules/manager/maven-wrapper/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/artifacts.spec.ts:17]:
```suggestion
process.env.CONTAINERBASE = 'true';

```

https://github.com/renovatebot/renovate/blob/ff648e687df564449241fb39716c5a514905b55f/lib/modules/manager/bundler/artifacts.spec.ts#L34

this shold fix the missing install command for `binarySource=install`

File: lib/modules/manager/maven-wrapper/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/maven-wrapper/artifacts.spec.ts:279]:
```suggestion
      { cmd: 'install-tool java 17.0.0' },
      {
```

see other comment

**Review by viceice — APPROVED** [review]

**Pull request merged**
